tmt/steps/discover/__init__.py

```diff
@@ -192,6 +195,68 @@ def save(self) -> None:
 
         self.write(Path('tests.yaml'), tmt.utils.dict_to_yaml(raw_test_data))
 
+    def _filter_for_rerun(self) -> None:
+        """ Filter out passed tests from previous run data """
+        assert isinstance(self.parent, tmt.base.Plan)  # narrow type
+        old_results: Path = self.parent.last_run_execute / 'results.yaml'
+        results = [
+            Result.from_serialized(data) for data in
+            tmt.utils.yaml_to_list(self.read(old_results))]
+        force_rerun_tests: list[str] = self.opt('force_rerun_test', [])
+        force_rerun_filters: list[str] = self.opt('force_rerun_filters', [])
+
+        results_failed: list[Result] = []
+        results_passed: list[Result] = []
+        for result in results:
+            if (
+                    force_rerun_tests and
+                    any(re.search(test, result.name) for test in force_rerun_tests)):
+                results_failed.append(result)
+                continue
+            if (
+                    result.result is not tmt.result.ResultOutcome.PASS and
+                    result.result is not tmt.result.ResultOutcome.INFO):
+                results_failed.append(result)
+            else:
+                results_passed.append(result)
+
+        # Filter out failed tests based on test name and serial number
+        filtered_tests: dict[str, list[tmt.base.Test]] = {}
+        for phase in self._tests:
+            current_phase_filtered: list[tmt.base.Test] = []
+            for test in self._tests[phase]:
+                # Check for any tests marked to rerun based on filter
+                if (
+                        force_rerun_filters and
+                        all(
+                            fmf_filter(rerun_filter, test._metadata, regexp=True)
+                            for rerun_filter in force_rerun_filters)):
+                    current_phase_filtered.append(test)
+                    # We need to also drop the test from passed so we don't have duplicate result
+                    for result_index in range(len(results_passed)):
+                        if (
+                                test.name == results_passed[result_index].name and
+                                test.serial_number == results_passed[result_index].serial_number):
+                            del results_passed[result_index]
+                            break
+                    continue
+                for result in results_failed:
+                    if test.name == result.name and test.serial_number == result.serial_number:
+                        current_phase_filtered.append(test)
+                        break
+            filtered_tests[phase] = current_phase_filtered
+        self._tests = filtered_tests
```

Comment:
Could you refactor this to reduce nesting and complexity, for example by extracting conditional checks into separate functions if necessary, and clarify the logic flow??

Reply:
Resolving, this was refactored.